torch/_custom_op/impl.py

```diff
@@ -417,6 +415,87 @@ def error_on_ctx():
 
         return inner
 
+    # Should be lazily registered when both backward and save_for_backward
+    # have been provided.
+    def _register_autograd_kernel(self):
+        assert self._has_impl("backward")
+        assert self._has_impl("save_for_backward")
+        kernel = construct_autograd_kernel(
+            self._schema,
+            self._output_differentiability,
+            self,
+            self._get_impl("save_for_backward").func,
+            self._get_impl("backward").func)
+        self._register_impl("autograd", kernel)
+
+    def impl_save_for_backward(self):
+        r"""Register a function that tells us what to save for backward.
+
+        Please see impl_backward for more details.
+        """
+        def inner(f):
+            self._register_impl("save_for_backward", f)
+            if self._has_impl("backward"):
+                self._register_autograd_kernel()
+        return inner
+
+    def impl_backward(self, output_differentiability=None):
+        r"""Registers a backward formula.
+
+        In order for the CustomOp to work with autograd, you need to register
+        a backward formula. There are two pieces to this:
+        1. You must give us a function to specify what to save for backward.
+           Call this the "save for backward" function.
+        2. You must give us a function that computes gradients. Call this the
+           "backward" function.
+
+        Use `impl_save_for_backward` to define a "save for backward" function
+        that specifies what gets saved for backward. The function should accept
+        two arguments ``(inputs, output)`` and return the quantities to be saved
+        for backward.
+
+        During runtime, when you call the CustomOp, PyTorch will invoke the
+        "save for backward" function with the inputs and output of the CustomOp.
+
+        Use `impl_backward` to define the "backward" function. The backward
+        function must accept ``(ctx, saved, *grads)``:
+        - ``ctx`` is a context object where we may provide information
+        - ``saved`` is exactly what gets returned from the "save for backward"
+          function
+        - ``grads`` is one or more gradients. The number of gradients matches
+          the number of outputs of the CustomOp.
+
+        The backward function must return a dict that maps the name of
+        an input to the CustomOp to its corresponding gradient. All inputs that
+        were declared to be Tensors in the CustomOp definition must be accounted
+        for in the dict. The gradient may be a Tensor or None.
```

Comment:
I'm not sure what's the benefit of using a dict here vs a tuple where there should be a 1-1 match anymore?
Is it just so that we don't need to pad with Nones anymore?

In the original design we were using dicts here for the registration API, it was a nice way to specify what each function was computing. But if this is doing eager evaluation is it still needed?

Reply:
> I'm not sure what's the benefit of using a dict here vs a tuple where there should be a 1-1 match anymore?
Is it just so that we don't need to pad with Nones anymore?

The main motivation is so that we don't need to pad with Nones anymore. Padding with Nones was a bit annoying to do, especially with autograd.Function that had a long list of inputs. I can be convinced to go back to the 1:1 match.

> In the original design we were using dicts here for the registration API, it was a nice way to specify what each function was computing. But if this is doing eager evaluation is it still needed?

We can still have a separate derivatives.yaml-like API for this, since there are separate technical benefits to the separate derivatives.yaml declarations vs the autograd.Function style. That one can look something like:

```py
myop.def_autograd({
   'x': ...
   'y': ...
})

```

It may be confusing to use dicts in both places, though.

Reply:
> We can still have a separate derivatives.yaml-like API for this

I was not suggesting that no. I agree we shouldn't mix them.

> The main motivation is so that we don't need to pad with Nones anymore.

Sounds fair!
We can always add tuple support if some people prefer later on.